File: src/test/java/plus/maa/backend/service/CopilotServiceTest.java
package plus.maa.backend.service;

import org.junit.jupiter.api.Test;
import plus.maa.backend.repository.entity.Copilot;
import plus.maa.backend.repository.entity.CopilotRating;
import reactor.util.function.Tuple2;
import reactor.util.function.Tuples;

import java.time.LocalDateTime;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertTrue;

public class CopilotServiceTest {

    @Test
    void testHotScores() {
        LocalDateTime now = LocalDateTime.now();
        LocalDateTime beforeWeek = now.minusDays(8L);

        // 一月前的作业，评分高，但是只有一条近期好评，浏览量高
        Copilot oldGreat = new Copilot();
        oldGreat.setUploadTime(beforeWeek.minusDays(14));
        oldGreat.setViews(20000L);
        CopilotRating rating1 = new CopilotRating();
        rating1.setRatingUsers(List.of(
                new CopilotRating.RatingUser("0", "Like", now),
                new CopilotRating.RatingUser("a", "Like", beforeWeek),
                new CopilotRating.RatingUser("b", "Like", beforeWeek),
                new CopilotRating.RatingUser("c", "Like", beforeWeek)));

        // 近期作业，含有差评，但是均为近期评分
        Copilot newGreat = new Copilot();
        newGreat.setUploadTime(now);
        newGreat.setViews(1000L);
        CopilotRating rating2 = new CopilotRating();
        rating2.setRatingUsers(List.of(new CopilotRating.RatingUser("d", "Like", now),
                new CopilotRating.RatingUser("e", "Like", now),
                new CopilotRating.RatingUser("a", "Like", now),
                new CopilotRating.RatingUser("f", "Like", now),
                new CopilotRating.RatingUser("g", "Like", now),
                new CopilotRating.RatingUser("b", "Dislike", now),
                new CopilotRating.RatingUser("c", "Like", now)));

        // 近期作业，差评较多，均为近期评分
        Copilot newBad = new Copilot();
        newBad.setUploadTime(now);
        newBad.setViews(500L);
        CopilotRating rating3 = new CopilotRating();
        rating3.setRatingUsers(List.of(new CopilotRating.RatingUser("a", "Like", now),
                new CopilotRating.RatingUser("b", "Dislike", now),
                new CopilotRating.RatingUser("c", "Dislike", now),
                new CopilotRating.RatingUser("d", "Dislike", now),
                new CopilotRating.RatingUser("e", "Dislike", now),
                new CopilotRating.RatingUser("f", "Like", now)));

        // 一月前的作业，评分高，但是只有一条近期好评，浏览量尚可
        Copilot oldNormal = new Copilot();
        oldNormal.setUploadTime(beforeWeek.minusDays(21L));
        oldNormal.setViews(4000L);
        CopilotRating rating4 = new CopilotRating();
        rating4.setRatingUsers(List.of(
                new CopilotRating.RatingUser("a", "Like", now),
                new CopilotRating.RatingUser("b", "Like", beforeWeek),
                new CopilotRating.RatingUser("c", "Like", beforeWeek)));

        // 新增作业，暂无评分
        Copilot newEmpty = new Copilot();
        newEmpty.setUploadTime(now);
        newEmpty.setViews(100L);
        CopilotRating rating5 = new CopilotRating();
        rating5.setRatingUsers(List.of());

        List<Tuple2<Copilot, CopilotRating>> orderRatings = List.of(Tuples.of(oldGreat, rating1),
                Tuples.of(newGreat, rating2), Tuples.of(newBad, rating3), Tuples.of(oldNormal, rating4),
                Tuples.of(newEmpty, rating5));

        for (Tuple2<Copilot, CopilotRating> orderRating : orderRatings) {
            orderRating.getT1().setHotScore(CopilotService.getHotScore(orderRating.getT1(), orderRating.getT2()));
        }

        // 近期好评 > 远古好评 > 近期新增 > 近期差评 > 远古一般
        assertTrue(newGreat.getHotScore() > oldGreat.getHotScore());
        assertTrue(newEmpty.getHotScore() > oldGreat.getHotScore());
        assertTrue(oldGreat.getHotScore() > newBad.getHotScore());
        assertTrue(oldNormal.getHotScore() > newBad.getHotScore());
    }

}

File: src/test/kotlin/plus/maa/backend/service/CopilotServiceTest.kt
package plus.maa.backend.service

import org.junit.jupiter.api.Test
import plus.maa.backend.repository.entity.Copilot
import plus.maa.backend.service.CopilotService.Companion.getHotScore
import java.time.LocalDateTime

class CopilotServiceTest {
    @Test
    fun testHotScores() {
        val now = LocalDateTime.now()
        val beforeWeek = now.minusDays(8L)
        val copilots = arrayOfNulls<Copilot>(5)
        val lastWeekLikeCounts = LongArray(5)
        val lastWeekDislikeCounts = LongArray(5)
        // 一月前的作业，评分高，但是只有一条近期好评，浏览量高
        val oldGreat = Copilot(doc = Copilot.Doc(title = "test"))
        oldGreat.uploadTime = beforeWeek.minusDays(14)
        oldGreat.views = 20000L
        copilots[0] = oldGreat
        lastWeekLikeCounts[0] = 1L
        lastWeekDislikeCounts[0] = 0L

        // 近期作业，含有差评，但是均为近期评分
        val newGreat = Copilot(doc = Copilot.Doc(title = "test"))
        newGreat.uploadTime = now
        newGreat.views = 1000L
        copilots[1] = newGreat
        lastWeekLikeCounts[1] = 6L
        lastWeekDislikeCounts[1] = 1L

        // 近期作业，差评较多，均为近期评分
        val newBad = Copilot(doc = Copilot.Doc(title = "test"))
        newBad.uploadTime = now
        newBad.views = 500L
        copilots[2] = newBad
        lastWeekLikeCounts[2] = 2L
        lastWeekDislikeCounts[2] = 4L

        // 一月前的作业，评分高，但是只有一条近期好评，浏览量尚可
        val oldNormal = Copilot(doc = Copilot.Doc(title = "test"))
        oldNormal.uploadTime = beforeWeek.minusDays(21L)
        oldNormal.views = 4000L
        copilots[3] = oldNormal
        lastWeekLikeCounts[3] = 1L
        lastWeekDislikeCounts[3] = 0L

        // 新增作业，暂无评分
        val newEmpty = Copilot(doc = Copilot.Doc(title = "test"))
        newEmpty.uploadTime = now
        newEmpty.views = 100L
        copilots[4] = newEmpty
        lastWeekLikeCounts[4] = 0L
        lastWeekDislikeCounts[4] = 0L

        for (i in 0..4) {
            copilots[i]!!.hotScore = getHotScore(copilots[i]!!, lastWeekLikeCounts[i], lastWeekDislikeCounts[i])
        }

        // 近期好评 > 远古好评 > 近期新增 > 近期差评 > 远古一般
        check(newGreat.hotScore > oldGreat.hotScore)
        check(newEmpty.hotScore > oldGreat.hotScore)
        check(oldGreat.hotScore > newBad.hotScore)
        check(oldNormal.hotScore > newBad.hotScore)
    }
}
